[v10.x] tls: expose keylog event on TLSSocket

mildsunrise opened the pull request:
Backport of #27654.

I had to modify `test-tls-keylog-tlsv13.js` to use TLS 1.2 connections as (AFAIK) v10 is not going to support TLS 1.3. I also modified the count from 5 to 1, because TLS 1.2 handshakes only emit 1 keylog event. Should I rename the file to `test-tls-keylog-tlsv12.js`?

mildsunrise commented:
wait, I see there's `ubuntu1804_sharedlibs_openssl110_x64`... do we need to support OpenSSL 1.1.0? if so then this can't be merged

bnoordhuis commented:
It's okay to disable the functionality with `#if OPENSSL_VERSION_NUMBER < 0x1010100fL`. That buildbot is so we know when a change breaks distro builds.

mildsunrise commented:
Okay! I've suppressed the `SSL_CTX_set_keylog_callback` call, the build should work now.

bnoordhuis commented:
@mildsunrise Do you want me to nominate you for collaborator status? I.e., get your commit bit? You've done enough high-impact work to qualify, IMO.

mildsunrise commented:
Oh, I forgot to skip the test when needed... it should be complete now

> @mildsunrise Do you want me to nominate you for collaborator status? I.e., get your commit bit? You've done enough high-impact work to qualify, IMO.

It'd be a pleasure! :)

bnoordhuis commented:
#32014 :)

BethGriggs commented:
In the most recent CI, the Windows failures are flakes, but the failure on ubuntu1804_sharedlibs_openssl110_x64 seems real - https://ci.nodejs.org/job/node-test-commit-linux-containered/18601/nodes=ubuntu1804_sharedlibs_openssl110_x64/testReport/junit/(root)/test/parallel_test_tls_keylog_tlsv13/

sam-github commented on test/parallel/test-tls-keylog-tlsv13.js:
```diff
@@ -0,0 +1,34 @@
+const common = require('../common');
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+if (/^(0\.|1\.0\.|1\.0\.0)/.test(process.versions.openssl))
```
```suggestion
if (/^(0\.|1\.0\.|1\.1\.0)/.test(process.versions.openssl))
```

^--- typo in third part of regex

sam-github replied:
I applied the change to unblock this, and I'll kick off ci.

sam-github commented on test/parallel/test-tls-keylog-tlsv13.js:
```diff
@@ -0,0 +1,34 @@
+  // Amount of keylog events depends on negotiated protocol
+  // version, so force a specific one:
+  minVersion: 'TLSv1.2',
+  maxVersion: 'TLSv1.2',
```
given this, the filename should be changed to remove "v13"

sam-github replied:
As I see you asked about now that I've read the whole review history...

sam-github commented:
Keylog support isn't available on that openssl version, but the regex that was added to check had a small typo. Passed for me locally after (by skipping the test).

sam-github commented:
> Should I rename the file to test-tls-keylog-tlsv12.js

Yes, please.

mildsunrise commented:
Totally right, that was a typo. I've renamed the file and squashed, hopefully everything is in order now (sorry for taking so many attempts to get it right 😅)

BethGriggs commented:
Landed in a2b0e9ef6a :tada:
